src/fw/drivers/imu/mmc5603nj/mmc5603nj.c

```diff
@@ -16,26 +16,334 @@
+
+static bool prv_mmc5603nj_init(void) {
+  if (s_initialized) {
+    return true;
+  }
+
+  if (!prv_mmc5603nj_check_whoami()) {
+    PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: WHO_AM_I check failed. Wrong device?");
+    return false;
+  }
+
+  // Reset the device
+  if (!prv_mmc5603nj_reset()) {
+    PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to reset");
+    return false;
+  }
+
+  s_initialized = true;
+  return true;
+}
+
+// Ask the compass for a 8-bit value that's programmed into the IC at the
+// factory. Useful as a sanity check to make sure everything came up properly.
+bool prv_mmc5603nj_check_whoami(void) {
+  uint8_t whoami = 0;
+  if (!prv_mmc5603nj_read(MMC5603NJ_REG_WHO_AM_I, 1, &whoami)) {
     return false;
+  }
+  return (whoami == MMC5603NJ_WHO_AM_I_VALUE);
+}
+
+static bool prv_mmc5603nj_reset(void) {
+  // Software reset
+  if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL1,
+                           MMC5603NJ_CTRL1_BANDWIDTH_6ms6 | MMC5603NJ_CTRL1_SW_RESET)) {
+    PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to reset device.");
+    return false;
+  }
+  psleep(MMC5603NJ_SW_RESET_DELAY_MS);
+
+  // Set operation
+  if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL0, MMC5603NJ_CTRL0_DO_SET)) {
+    PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to set coils.");
+    return false;
+  }
+  psleep(MMC5603NJ_SET_DELAY_MS);
+
+  // Reset operation
+  if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL0, MMC5603NJ_CTRL0_DO_RESET)) {
+    PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to reset coils.");
+    return false;
+  }
+  psleep(MMC5603NJ_SET_DELAY_MS);
+  return true;
+}
+
+// Configure ODR
+
+bool prv_mmc5603nj_set_sample_rate_hz(uint8_t rate_hz) {
+  if (rate_hz == s_sample_rate_hz) {
+    return true;
+  }
+
+  PBL_LOG(LOG_LEVEL_DEBUG, "MMC5603NJ: Setting sample rate to %d Hz", rate_hz);
+
+  // Reset device runtime status (disabling continuous mode/etc)
+  if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL2, 0x00)) {
+    return false;
+  }
+
+  // Do one final read to reset any data ready flags
+  MagData discard_sample;
+  prv_mmc5603nj_get_sample(&discard_sample);
+
+  if (rate_hz > 0) {
+    // Set new sampling rate
+    if (!prv_mmc5603nj_write(MMC5603NJ_REG_ODR, rate_hz)) {
+      PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to update ODR.");
+    }
+
+    // Retrigger calculation of measurements rates
+    if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL0,
+                             MMC5603NJ_CTRL0_AUTO_SR_EN | MMC5603NJ_CTRL0_CMM_FREQ_EN)) {
+      PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to trigger measurement calculation update.");
+    }
+
+    // Start continuous mode
+    if (!prv_mmc5603nj_write(MMC5603NJ_REG_CTRL2, MMC5603NJ_CTRL2_AUTOSET_PRD_100 |
+                                                      MMC5603NJ_CTRL2_PRD_SET_EN |
+                                                      MMC5603NJ_CTRL2_CMM_EN)) {
+      PBL_LOG(LOG_LEVEL_ERROR, "MMC5603NJ: Failed to start continuous mode.");
+      mutex_unlock(s_mag_mutex);
```

Comment:
not locking in this function

Reply:
Whoops. Thanks for rooting up this vestigial unlock!